templates/cloudflare-workers/worker/TldrawDurableObject.ts

```diff
@@ -0,0 +1,134 @@
+	private r2: R2Bucket
+	// the room ID will be missing whilst the room is being initialized
+	private roomId: string | null = null
+	// when we load the room from the R2 bucket, we keep it here. it's a promise so we only ever
+	// load it once.
+	private roomPromise: Promise<TLSocketRoom<TLRecord, void>> | null = null
+
+	constructor(state: DurableObjectState, env: Environment) {
+		super(state, env)
+		this.r2 = env.TLDRAW_BUCKET
+
+		state.blockConcurrencyWhile(async () => {
+			this.roomId = ((await this.ctx.storage.get('roomId')) ?? null) as string | null
+		})
+	}
+
+	private readonly router = AutoRouter({
+		catch: (e) => {
+			console.log(e)
+			return error(e)
+		},
+	})
+		// when we get a connection request, we stash the room id if needed and handle the connection
+		.get('/connect/:roomId', async (request) => {
+			if (!this.roomId) {
+				await this.ctx.blockConcurrencyWhile(async () => {
+					await this.ctx.storage.put('roomId', request.params.roomId)
+					this.roomId = request.params.roomId
+				})
+			}
+			return this.handleConnect(request)
+		})
+
+	// `fetch` is the entry point for all requests to the Durable Object
+	fetch(request: Request): Response | Promise<Response> {
+		return this.router.fetch(request)
+	}
+
+	// what happens when someone tries to connect to this room?
+	async handleConnect(request: IRequest): Promise<Response> {
+		// extract query params from request, should include instanceId
+		const sessionKey = request.query.sessionKey as string
+		if (!sessionKey) return error(400, 'Missing sessionKey')
+
+		// Create the websocket pair for the client
+		const { 0: clientWebSocket, 1: serverWebSocket } = new WebSocketPair()
+		serverWebSocket.accept()
+
+		// load the room, or retrieve it if it's already loaded
+		const room = await this.getRoom()
+
+		// connect the client to the room
+		room.handleSocketConnect(sessionKey, serverWebSocket)
+
+		// return the websocket connection to the client
+		return new Response(null, { status: 101, webSocket: clientWebSocket })
+	}
+
+	getRoom() {
+		const roomId = this.roomId
+		if (!roomId) throw new Error('Missing roomId')
+
+		if (!this.roomPromise) {
+			this.roomPromise = (async () => {
+				// fetch the room from R2
+				const roomFromBucket = await this.r2.get(`rooms/${roomId}`)
+
+				// if it doesn't exist, we'll just create a new empty room
+				const initialSnapshot = roomFromBucket
+					? ((await roomFromBucket.json()) as RoomSnapshot)
+					: undefined
+
+				// create a new TLSocketRoom. This handles all the sync protocol & websocket connections.
+				// it's up to us to persist the room state to R2 when needed though.
+				return new TLSocketRoom<TLRecord, void>({
+					schema,
+					initialSnapshot,
+					onSessionRemoved: async (_room, args) => {
+						// persist whenever the number of active sessions drops to zero
+						if (args.numSessionsRemaining > 0) return
+						await this.schedulePersistToR2()
+					},
```

Comment:
I don't think this is actually necessary. onDataChange will have been called if the room data changed so there's no need to do it again here.

```suggestion
```